chiselFrontend/src/main/scala/chisel3/core/MultiClock.scala

```diff
@@ -0,0 +1,25 @@
+// See LICENSE for license details.
+
+package chisel3.core
+
+import scala.language.experimental.macros
+
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.{SourceInfo}
+
+private[chisel3] final case class ClockAndReset(clock: Clock, reset: Bool)
+
+object withClockAndReset {  // scalastyle:ignore object.name
+  def apply[T](clock: Clock, reset: Bool)(block: => T)(implicit sourceInfo: SourceInfo): T = {
+    // Save parentScope
+    val parentScope = Builder.currentClockAndReset
+    Builder.currentClockAndReset = Some(ClockAndReset(clock, reset))
+    val res = block // execute block
+    // Return to old scope
+    Builder.currentClockAndReset = parentScope
+    res
```

Comment:
The example is such a compelling use case that I think we should just say "just don't do that" to @ducky64's example.

If it returns Unit, people will just use vars to achieve the same thing badly.